vendor: cAdvisor v0.39.0

bobbypage opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

/kind bug

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature
/kind design

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:

This PR updates cAdvisor to v0.39.0. For every release of k8s cAdvisor is updated to it's new minor version.  

For k8s 1.21, we'll be updating cAdvisor to v0.39.0

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #99739
Fixes #95039

#### Special notes for your reviewer:

cAdvisor v0.39.0 changelog: https://github.com/google/cadvisor/releases/tag/v0.39.0

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

dims commented:
just a couple of more steps

```
GO111MODULE=on go mod edit -dropreplace github.com/docker/go-events
GO111MODULE=on go mod edit -dropreplace github.com/gogo/googleapis
```

dims commented:
looks like this is shaping up well. thanks @bobbypage

bobbypage commented:
Thanks @dims! 

Trying it out as WIP on a latest commit of cadvisor before cutting v0.39.0 release. Looks like there is two issues that will need to be fixed:

```
# k8s.io/kubernetes/pkg/kubelet/cadvisor
pkg/kubelet/cadvisor/cadvisor_linux.go:112:23: not enough arguments in call to manager.New
	have (*memory.InMemoryCache, sysfs.SysFs, struct { Interval *time.Duration; AllowDynamic *bool }, container.MetricSet, *http.Client, []string, string)
	want (*memory.InMemoryCache, sysfs.SysFs, struct { Interval *time.Duration; AllowDynamic *bool }, container.MetricSet, *http.Client, []string, string, *metrics.DenyList)
# k8s.io/kubernetes/pkg/kubelet/dockershim/libdocker
pkg/kubelet/dockershim/libdocker/kube_docker_client.go:148:45: not enough arguments in call to d.client.ContainerCreate
	have (context.Context, *container.Config, *container.HostConfig, *network.NetworkingConfig, string)
	want (context.Context, *container.Config, *container.HostConfig, *network.NetworkingConfig, *"k8s.io/kubernetes/vendor/github.com/opencontainers/image-spec/specs-go/v1".Platform, string)
```

bobbypage commented:
Also note to self, steps to update:

```
CADVISOR_PIN="${SHA}"
hack/pin-dependency.sh github.com/google/cadvisor ${CADVISOR_PIN}
```

and this little script to update all pins / vendor:

```
while true; do
  PINS_NEEDED=$(hack/lint-dependencies.sh  | grep "hack/pin-dependency.sh" | grep -v "Use hack/pin-dependency.sh" | awk '{$1=$1;print}')

  if [[ -z "${PINS_NEEDED}" ]]; then
    echo ">>> all done; no more pins"
    break
  fi

  while IFS= read -r line
  do
    echo ">>> Running pin ${line}"
    eval "${line}"
  done <<< "${PINS_NEEDED}"

  echo ">>> Vendoring"
  hack/update-vendor.sh

  REMOVES_NEEDED=$(hack/lint-dependencies.sh  | grep 'dropreplace' | awk '{$1=$1;print}')
  while IFS= read -r line
  do
    echo ">>> Running removal ${line}"
    eval "${line}"
  done <<< "${REMOVES_NEEDED}"
done
```

bobbypage commented:
For the two errors:

```
pkg/kubelet/cadvisor/cadvisor_linux.go:112:23: not enough arguments in call to manager.New
	have (*memory.InMemoryCache, sysfs.SysFs, struct { Interval *time.Duration; AllowDynamic *bool }, container.MetricSet, *http.Client, []string, string)
	want (*memory.InMemoryCache, sysfs.SysFs, struct { Interval *time.Duration; AllowDynamic *bool }, container.MetricSet, *http.Client, []string, string, *metrics.DenyList)
```

cAdvisor deny list argument missed caused by change in cAdvisor.New() due to https://github.com/google/cadvisor/pull/2774

```
pkg/kubelet/dockershim/libdocker/kube_docker_client.go:148:45: not enough arguments in call to d.client.ContainerCreate
	have (context.Context, *container.Config, *container.HostConfig, *network.NetworkingConfig, string)
	want (context.Context, *container.Config, *container.HostConfig, *network.NetworkingConfig, *"k8s.io/kubernetes/vendor/github.com/opencontainers/image-spec/specs-go/v1".Platform, string)
```

caused by docker update, see https://github.com/moby/moby/pull/40725 

I think something like changes below will be sufficient:

```
diff --git a/pkg/kubelet/cadvisor/cadvisor_linux.go b/pkg/kubelet/cadvisor/cadvisor_linux.go
index 0c701cb98d3..befbebeddab 100644
--- a/pkg/kubelet/cadvisor/cadvisor_linux.go
+++ b/pkg/kubelet/cadvisor/cadvisor_linux.go
@@ -43,6 +43,7 @@ import (
        cadvisorapi "github.com/google/cadvisor/info/v1"
        cadvisorapiv2 "github.com/google/cadvisor/info/v2"
        "github.com/google/cadvisor/manager"
+       cadvisordenymetrics "github.com/google/cadvisor/metrics"
        "github.com/google/cadvisor/utils/sysfs"
        utilfeature "k8s.io/apiserver/pkg/util/feature"
        "k8s.io/klog/v2"
@@ -109,7 +110,11 @@ func New(imageFsInfoProvider ImageFsInfoProvider, rootPath string, cgroupRoots [
        }

        // Create the cAdvisor container manager.
-       m, err := manager.New(memory.New(statsCacheDuration, nil), sysFs, housekeepingConfig, includedMetrics, http.DefaultClient, cgroupRoots, "")
+       denyList, err := cadvisordenymetrics.NewDenyList([]string{})
+       if err != nil {
+               return nil, err
+       }
+       m, err := manager.New(memory.New(statsCacheDuration, nil), sysFs, housekeepingConfig, includedMetrics, http.DefaultClient, cgroupRoots, "", denyList)
        if err != nil {
                return nil, err
        }
diff --git a/pkg/kubelet/dockershim/libdocker/kube_docker_client.go b/pkg/kubelet/dockershim/libdocker/kube_docker_client.go
index 81b6582e2bc..3dfe4cf0145 100644
--- a/pkg/kubelet/dockershim/libdocker/kube_docker_client.go
+++ b/pkg/kubelet/dockershim/libdocker/kube_docker_client.go
@@ -145,7 +145,7 @@ func (d *kubeDockerClient) CreateContainer(opts dockertypes.ContainerCreateConfi
        if opts.HostConfig != nil && opts.HostConfig.ShmSize <= 0 {
                opts.HostConfig.ShmSize = defaultShmSize
        }
-       createResp, err := d.client.ContainerCreate(ctx, opts.Config, opts.HostConfig, opts.NetworkingConfig, opts.Name)
+       createResp, err := d.client.ContainerCreate(ctx, opts.Config, opts.HostConfig, opts.NetworkingConfig, nil, opts.Name)
        if ctxErr := contextError(ctx); ctxErr != nil {
                return nil, ctxErr
        }
```

bobbypage commented:
And runc changes:

```
# k8s.io/kubernetes/pkg/kubelet/cm
pkg/kubelet/cm/cgroup_manager_linux.go:476:17: undefined: configs.Wildcard
pkg/kubelet/cm/cgroup_manager_linux.go:477:17: undefined: configs.Wildcard
pkg/kubelet/cm/cgroup_manager_linux.go:507:18: undefined: configs.Wildcard
pkg/kubelet/cm/cgroup_manager_linux.go:508:18: undefined: configs.Wildcard
pkg/kubelet/cm/container_manager_linux.go:400:19: undefined: configs.Wildcard
pkg/kubelet/cm/container_manager_linux.go:401:19: undefined: configs.Wildcard
```

bobbypage committed:
Fixes after runc libcontainer and docker update

- libcontainer renamed
  `github.com/opencontainers/runc/libcontainer/configs` to
  `github.com/opencontainers/runc/libcontainer/devices` so use the new
  references

- Update `dockershim` `ContainerCreate` call after docker update to
  v20.10.2

bobbypage commented:
[cAdvisor v0.39.0](https://github.com/google/cadvisor/releases/tag/v0.39.0) has been cut. Pinned latest cAdvisor v0.39.0 tag, re-vendored, and updated the PR.

bobbypage commented:
/kind bug

bobbypage commented:
/priority important-soon
/triage accepted

bobbypage commented:
@kubernetes/milestone-maintainers can a milestone maintainer stamp a milestone v1.21 on this PR please? Thanks! :)

bobbypage commented:
/assign @dims

bobbypage commented:
/cc @iwankgb @dims @SergeyKanzhelev @mrunalp @dashpole @harche

harche commented:
/test pull-kubernetes-node-crio-e2e

harche commented:
/lgtm

liggitt commented on vendor/github.com/containerd/containerd/platforms/cpuinfo.go:
```diff
@@ -0,0 +1,122 @@
+
+func init() {
+	if isArmArch(runtime.GOARCH) {
+		cpuVariant = getCPUVariant()
```
this makes calls that can return errors and log... is that expected on startup for the kubelet (it could happen when running things like `kubelet --help`)?

bobbypage replied:
Thanks for looking through the dep changes.

With regards to this `init()` function, looks like this was refactored in https://github.com/containerd/containerd/pull/4953 and moved out of init(), but the change isn't in containerd v1.4.4 (only in unreleased containerd v1.5).

It looks like the main place error can happen here is if it can't open `/proc/cpuinfo` (https://github.com/containerd/containerd/blob/release%2F1.4/platforms/cpuinfo.go#L49-L53) which is hopefully unlikely :). Not clear what we can do here unfortunately considering we can't move to containerd v1.5 yet...

neolit123 replied:
> It looks like the main place error can happen here is if it can't open /proc/cpuinfo (https://github.com/containerd/containerd/blob/release%2F1.4/platforms/cpuinfo.go#L49-L53) which is hopefully unlikely

if this code path happens on Windows, there is no /proc/ there.

neolit123 replied:
my comment was more of a remark that code should be written as preemptively OS-portable.

k8s doesn't produce ARM binaries for Windows. by the time (and if) we start doing that, the containerd version will likely be bumped to not have this init() code.

liggitt reviewed (commented):
* dependency mechanics lgtm
* deltas in docker/docker lgtm
* swept for new init() functions and found one that looked a little weird

would like an approve of the cilium and runc updates from sig-node-approvers

bobbypage committed:
vendor: cadvisor v0.39.0

Main upgrades:
- github.com/opencontainers/runc v1.0.0-rc93
- github.com/containerd/containerd v1.4.4
- github.com/docker/docker v20.10.2
- github.com/mrunalp/fileutils v0.5.0
- github.com/opencontainers/selinux v1.8.0
- github.com/cilium/ebpf v0.2.0

bobbypage commented:
Rebased.

harche commented:
/test pull-kubernetes-node-crio-e2e

bobbypage commented:
@mrunalp can you PTAL at the runc / cilium changes? Thanks!

dims reviewed (approved):
/approve
/lgtm

dims commented:
/assign @liggitt

liggitt commented:
/approve
/hold for node lgtm - https://github.com/kubernetes/kubernetes/pull/99875#issuecomment-793488526

mrunalp commented:
/lgtm

liggitt commented:
/hold cancel

jc42jc commented:
Hello,
The fix didn't fix the issue. I still have this error on Kubernetes / K3S 1.21.

dims commented:
@jc42jc can you please open an issue in k3s repo and ask them to debug/triage?
